### PR TITLE
Hck 3564 alter script for comment modification

### PR DESCRIPTION
**Pull request by ViktorMasnyi**
list of implemented
- remove recursive document tree traverse, while getting properties from the persistence schema
- refactor to return DTO from every script statement, refactor statements sorting function

### CODE REVIEW

File: forward_engineering/helpers/updateHelpers/viewHelper.js [hunk 1]
```diff
@@ -1,17 +1,11 @@
 const { getViewScript: generateViewScript, getOptionsScript } = require('../viewHelper');
 const { tab } = require('../generalHelper');
 const { dependencies } = require('../appDependencies');
+const { AlterScriptDto } = require("../types/AlterScriptDto");
 let _;
```
Comment (ArzamastsevVladyslav) [inline forward_engineering/helpers/updateHelpers/viewHelper.js:5]: Find a way to get rid of caching _ dependency. Either pass it as functions' dependencies or use globall dependency object

File: forward_engineering/helpers/updateHelpers/udtHelper.js [hunk 2]
```diff
@@ -24,24 +25,31 @@ const getRenameType = (renameData) =>
...
 	RENAME "${renameData.oldFieldName}" TO "${renameData.newFieldName}";`;
 
-const getDropUDT = (dropUDTData) => ([{
-	...scriptData,
-	script: `DROP TYPE IF EXISTS "${dropUDTData.keyspaceName}"."${dropUDTData.typeName}";`,
-	deleted: true,
-}]);
+const getDropUDT = (dropUDTData) => ([
+	AlterScriptDto.getInstance(
+		[`DROP TYPE IF EXISTS "${dropUDTData.keyspaceName}"."${dropUDTData.typeName}";`],
```
Comment (ArzamastsevVladyslav) [inline forward_engineering/helpers/updateHelpers/udtHelper.js:30]: DDL templates are supposed to be in templates.js file. ddlProvider must've been used for filling templates with values

File: forward_engineering/helpers/updateHelper.js [hunk 3]
```diff
@@ -402,72 +413,61 @@ const generateScript = (child, udtMap, data, column, mode) => {
...
+    const deletedModelDefinitions = sortAddedUdtResult?.properties?.deleted;
+	
+    const addedModelDefinitionsScripts = generateScript(addedModelDefinitions.items, udtMap, data, 'udt', 'add');
+    const modifiedModelDefinitionsScripts = generateScript(modifiedModelDefinitions.items, udtMap, data, 'udt', 'update');
+    const deletedModelDefinitionsScript = generateScript(deletedModelDefinitions.items, udtMap, data, 'udt', 'delete');
+	
+	return [
+			...modifiedEntitiesScripts,
+			...addedEntitiesScripts,
+			...modifiedEntitiesScript,
```
Comment (ArzamastsevVladyslav) [inline forward_engineering/helpers/updateHelper.js:460]: Create separate functions that handle generation of scripts related to entities, containers, views etc

File: forward_engineering/api.js [hunk 4]
```diff
@@ -232,4 +232,4 @@ module.exports = {
 
 const getScript = (structure) => {
```
Comment (ArzamastsevVladyslav) [inline forward_engineering/api.js:233]: api.js 's function should invoke alterScriptBuilder's functions for building the alter script and checking if it contains 'DROP' statements. I do not see these changes

File: forward_engineering/helpers/updateHelper.js [hunk 5]
```diff
@@ -489,28 +489,35 @@ const getCommentedDropScript = (scriptsData, data) => {
...
-			return scriptData;
+	return scriptsData.map((dto = {}) => {
+		if (!dto?.scripts[0]?.isDropScript || !dto?.scripts[0]?.script) {
+			return dto;
 		}
 		return {
-			...scriptData,
-			script: commentDeactivatedStatement(scriptData.script, false),
+			...dto,
+			scripts: dto.scripts.map(scriptObject => {
```
Comment (ArzamastsevVladyslav) [inline forward_engineering/helpers/updateHelper.js:498]: activation status is not taken into account in commenting out process

**Review by ArzamastsevVladyslav — CHANGES_REQUESTED** [review]

File: forward_engineering/helpers/updateHelper.js [hunk 6]
```diff
@@ -387,6 +395,9 @@ const columns = {
 }
 
 const generateScript = (child, udtMap, data, column, mode) => {
```
Comment (ArzamastsevVladyslav) [inline forward_engineering/helpers/updateHelper.js:397]: It is preferrable to name functions like <do-something-**_Script_**> if the return a script - i.e. a string. If it doesn't you should name it like <do-something-**_ScriptDto_**>

Comment (ArzamastsevVladyslav) [reply forward_engineering/helpers/updateHelper.js:397]: You should've also included JsDocs to refactored and added functions

**Review by ArzamastsevVladyslav — CHANGES_REQUESTED** [review]

File: forward_engineering/helpers/types/AlterScriptDto.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * @param scripts {Array<string>}
+     * @param isActivated {!boolean}
+     * @param isDropScripts {!boolean}
+     * @param isModifyScript {!boolean}
+     * @param isAddScript {!boolean}
+     * @param modelLevel {?string} 'field' | 'index' | 'udt'
+     * @return {AlterScriptDto | undefined}
+     * */
+    static getInstance(scripts, isActivated, isDropScripts, isModifyScript, isAddScript, modelLevel ) {
```
[the remark this answers is not in the document]
Comment (ArzamastsevVladyslav) [reply forward_engineering/helpers/types/AlterScriptDto.js:51]: And create a separate static method for that

File: forward_engineering/helpers/updateHelpers/viewHelper.js (hunk 1 above)
Comment (ViktorMasnyi) [reply forward_engineering/helpers/updateHelpers/viewHelper.js:5]: I propose not to modify this, as we expecting the update of the flow how we're working with plugin dependencies

Comment (ViktorMasnyi) [reply forward_engineering/helpers/updateHelpers/viewHelper.js]: got lodash from the global dependencies object

File: forward_engineering/helpers/updateHelper.js (hunk 3 above)
Comment (ViktorMasnyi) [reply forward_engineering/helpers/updateHelper.js]: done with this

File: forward_engineering/helpers/updateHelper.js (hunk 5 above)
Comment (ViktorMasnyi) [reply forward_engineering/helpers/updateHelper.js:552]: getCommentedDropScript function does this - I added the activate/deactivate logic there

**Review by ArzamastsevVladyslav — CHANGES_REQUESTED** [review]

File: forward_engineering/helpers/updateHelpers/tableHelper.js [hunk 8]
```diff
@@ -3,50 +3,55 @@ const { getNamesByIds } = require("../schemaHelper");
...
 	addTable: [],
 	deleteTable: [],
 };
 
-const setDependencies = ({ lodash }) => _ = lodash;
-
-const removeColumnStatement = columnName => `DROP "${columnName}";`;
-const addColumnStatement = columnData => `ADD "${columnData.name}" ${columnData.type}`;
-
 const alterTablePrefix = (tableName, keySpace) =>
```
[the remark this answers is not in the document]
Comment (ViktorMasnyi) [reply forward_engineering/helpers/updateHelpers/tableHelper.js:14]: done

File: forward_engineering/helpers/indexHelper.js [hunk 9]
```diff
@@ -233,10 +223,8 @@ const getSearchIndex = ({ tableName, columns, index, isActivated, ifNotExist })
 	const config = indexConfig ? `\n${tab(`AND CONFIG ${indexConfig}`)}` : '';
 	const indexOptions = getSearchIndexOptions(index.options);
 	const options = indexOptions ? `\n${tab(`AND OPTIONS ${indexOptions}`)}` : '';
-	return `CREATE SEARCH INDEX ${ifNotExist ? `IF NOT EXISTS `:``}ON ${tableName}\n` + 
-		`${tab(`WITH COLUMNS\n${tab(columnsStatement)}`)}` + 
-		profiles + config + options +
-		';';
+	
+	return dependencies.provider.createSearchIndex({ifNotExist, tableName, columnsStatement: tab(columnsStatement, 2), profiles, config, options});
```
[the remark this answers is not in the document]
Comment (ViktorMasnyi) [reply forward_engineering/helpers/indexHelper.js:227]: https://hackolade.atlassian.net/browse/HCK-3651

File: forward_engineering/helpers/alterScriptFromDeltaHelper.js [hunk 10]
```diff
@@ -401,165 +407,145 @@ const generateScript = (child, udtMap, data, column, mode) => {
...
+const isDropInStatements = (child, udtMap, data) => {
+	return [
+		...getEntitiesDto(child, udtMap, data),
+		...getContainersDto(child, udtMap, data),
+		...getViewsDto(child, udtMap, data),
+		...getModelDefinitionsDto(child, udtMap, data)
+	]
+		.filter(Boolean)
+		.flatMap(dto => dto.scripts)
+		.some(scriptData => !!scriptData.script && scriptData.isDropScript);
```
Comment (ArzamastsevVladyslav) [inline forward_engineering/helpers/alterScriptFromDeltaHelper.js:545]: Take activation into account, move to alterScriptBuilder

**Review by ArzamastsevVladyslav — CHANGES_REQUESTED** [review]

**Review by ArzamastsevVladyslav — APPROVED** [review]

**Pull request merged**